File: projectionist/library/watch_progress.py
"""Shared watch-progress classification for library titles.

Returns one of ``unwatched`` | ``partial`` | ``watched``.

Rules
-----
**Movies**
- ``watched``: ``view_count > 0``
- ``partial``: unfinished in-progress playhead (``view_offset_ms > 0``) with
  ``view_count == 0``. Requires Plex sync of ``view_offset_ms``; when that
  field is absent CuratorX can only distinguish watched vs unwatched.
- ``unwatched``: otherwise

**Shows**
- ``watched``: ``total_episode_count > 0`` and ``unwatched_episode_count == 0``
  (``unwatched_episode_count`` must be present — ``None`` is not zero)
- ``partial``: some but not all episodes watched
  (``0 < unwatched < total``)
- When ``total_episode_count > 0`` but ``unwatched_episode_count`` is missing
  (episode progress not synced), treat as ``unwatched`` rather than inferring
  "all watched".
- When episode totals are unavailable, fall back to show-level ``view_count``
  (watched if ``> 0``, else unwatched). No honest partial without episode data.
"""

from __future__ import annotations

from typing import Any, Literal, Mapping

WatchProgressState = Literal["unwatched", "partial", "watched"]


def _as_nonneg_int(value: Any) -> int:
    try:
        return max(0, int(value or 0))
    except (TypeError, ValueError):
        return 0


def watch_progress_state(item: Mapping[str, Any] | None) -> WatchProgressState:
    """Classify a library/query/title-card mapping into a watch-progress state."""
    if not item:
        return "unwatched"

    media = str(item.get("media_type") or "").strip().lower()
    view_count = _as_nonneg_int(item.get("view_count"))
    view_offset_ms = _as_nonneg_int(item.get("view_offset_ms"))
    total = _as_nonneg_int(item.get("total_episode_count"))
    raw_unwatched = item.get("unwatched_episode_count")

    # Episode totals imply show semantics even when media_type is omitted
    # (e.g. privacy derive_watch_state payloads).
    is_show = media in {"show", "tv", "series"} or total > 0
    if is_show:
        if total > 0:
            # Missing unwatched count means episode progress was never synced;
            # do not treat None as 0 (all watched).
            if raw_unwatched is None:
                return "unwatched"
            unwatched = _as_nonneg_int(raw_unwatched)
            if unwatched == 0:
                return "watched"
            if unwatched < total:
                return "partial"
            return "unwatched"
        if view_count > 0:
            return "watched"
        return "unwatched"

    # Movies and unknown media types use movie semantics.
    if view_count > 0:
        return "watched"
    if view_offset_ms > 0:
        return "partial"
    return "unwatched"

File: frontend/src/lib/watchProgress.js
/**
 * Shared watch-progress classification for poster badges.
 * Mirrors curatorx.library.watch_progress.watch_progress_state.
 *
 * Movies:
 * - watched: view_count > 0
 * - partial: view_offset_ms > 0 with view_count === 0 (needs Plex sync)
 * - unwatched: otherwise
 *
 * Shows:
 * - watched: total_episode_count > 0 && unwatched_episode_count === 0
 * - partial: 0 < unwatched < total
 * - unwatched when total > 0 but unwatched_episode_count is null (not synced)
 * - fallback when episode totals missing: view_count > 0 → watched
 */

function asNonNegInt(value) {
  const n = Number(value);
  if (!Number.isFinite(n) || n <= 0) return 0;
  return Math.floor(n);
}

/**
 * @param {Record<string, unknown> | null | undefined} item
 * @returns {"unwatched" | "partial" | "watched"}
 */
export function watchProgressState(item) {
  if (!item || typeof item !== "object") return "unwatched";

  // Privacy/member sanitization may replace raw counters with watch_state.
  const explicit = String(item.watch_state || "")
    .trim()
    .toLowerCase();
  if (explicit === "watched") return "watched";
  if (explicit === "partial" || explicit === "in_progress") return "partial";
  if (explicit === "unwatched") return "unwatched";

  const media = String(item.media_type || "")
    .trim()
    .toLowerCase();
  const viewCount = asNonNegInt(item.view_count);
  const viewOffsetMs = asNonNegInt(item.view_offset_ms);
  const total = asNonNegInt(item.total_episode_count);
  const rawUnwatched = item.unwatched_episode_count;

  // Episode totals imply show semantics even when media_type is omitted.
  const isShow = media === "show" || media === "tv" || media === "series" || total > 0;
  if (isShow) {
    if (total > 0) {
      // Missing unwatched count means episode progress was never synced.
      if (rawUnwatched == null) return "unwatched";
      const unwatched = asNonNegInt(rawUnwatched);
      if (unwatched === 0) return "watched";
      if (unwatched < total) return "partial";
      return "unwatched";
    }
    return viewCount > 0 ? "watched" : "unwatched";
  }

  if (viewCount > 0) return "watched";
  if (viewOffsetMs > 0) return "partial";
  return "unwatched";
}

/** Aria / title copy for a watch-progress state. */
export function watchProgressLabel(state) {
  if (state === "watched") return "Watched";
  if (state === "partial") return "In progress";
  return "";
}
